src/electionguard/async.hpp

```diff
@@ -0,0 +1,324 @@
+                if (isReady(task)) {
+                    results.push_back(std::move(task.get()));
+                }
+            }
+        } while (results.size() < tasks.size());
+        return results;
+    }
+
+    /// <summary>
+    /// A simple asynchronous thread safe queue using locks.
+    /// </summary>
+    template <typename T> class EG_INTERNAL_API AsyncQueue
+    {
+      public:
+        AsyncQueue() : _head(new Node), _tail(_head.get()) {}
+        AsyncQueue(const AsyncQueue &other) = delete;
+        AsyncQueue &operator=(const AsyncQueue &other) = delete;
+
+        bool empty()
+        {
+            std::lock_guard<std::mutex> lock(_head_mutex);
+            return (_head.get() == _getTail());
+        }
+
+        std::shared_ptr<T> pop()
+        {
+            const auto old = _tryPop();
+            return old ? old->data : std::shared_ptr<T>();
+        }
+
+        bool pop(T &value)
+        {
+            const auto old = _tryPop(value);
+            return old != nullptr;
+        }
+
+        void push(T value)
+        {
+            auto data = std::make_shared<T>(std::move(value));
+            auto node = std::unique_ptr<Node>(new Node);
+            {
+                std::lock_guard<std::mutex> lock(_tail_mutex);
+                _tail->data = data;
+                Node *const new_tail = node.get();
+                _tail->next = std::move(node);
+                _tail = new_tail;
+            }
+            _condition.notify_one();
+        }
+
+      private:
+        struct Node {
+            std::shared_ptr<T> data;
+            std::unique_ptr<Node> next;
+        };
+
+        std::unique_ptr<Node> _head;
+        Node *_tail;
+
+        std::mutex _head_mutex;
+        std::mutex _tail_mutex;
+
+        std::condition_variable _condition;
+
+        Node *_getTail()
+        {
+            std::lock_guard<std::mutex> lock(_tail_mutex);
+            return _tail;
+        }
+
+        std::unique_ptr<Node> _pop()
+        {
+            if (_head.get() == _getTail()) {
+                return nullptr;
+            }
+            auto old = std::move(_head);
+            _head = std::move(old->next);
+            return old;
+        }
+
+        std::unique_ptr<Node> _tryPop()
+        {
+            std::lock_guard<std::mutex> lock(_head_mutex);
+            if (_head.get() == _getTail()) {
+                return std::unique_ptr<Node>();
+            }
+            return _pop();
+        }
+
+        std::unique_ptr<Node> _tryPop(T &value)
+        {
+            std::lock_guard<std::mutex> lock(_head_mutex);
+            if (_head.get() == _getTail()) {
+                return std::unique_ptr<Node>();
+            }
+            value = std::move(*_head->data);
+            return _pop();
+        }
+    };
+
+    /// <summary>
+    /// A generic wrapper around a callable type such as a function.
+    /// </summary>
+    class EG_INTERNAL_API CallableWrapper
```

Comment:
to get things working properly i needed an object that is copy-construbtable but most of our api's deal with unique_ptr's so this just encapsulates the actual underlying callable function in a type that can be copied